0003-monitoring-system-end-to-end-tests/0003-monitoring-system-end-to-end-tests.md

```diff
@@ -0,0 +1,79 @@
+# RFC 0003 - Monitoring System End To End Tests
+
+We're currently building a monitoring system that supports a large number of
+Kubernetes clusters by provisioning Prometheus servers for each monitored
+cluster in a sharded architecture. We would like to have end-to-end tests for
+this system that ensure it functions correctly and to prevent regressions.
+
+## Summary of the Monitoring System
+
+The system is built initially with Giant Swarm architecture in mind where each
+installation has a _Management Cluster_ where monitoring and infrastructure
+operators live and a number of _Workload Clusters_ where customers' workloads
+live.
+
+The system consists of [prometheus-meta-operator][] which reacts to clusters
+being created/deleted either via [Cluster-API][] or GS-specific API by
+provisioning/maintaining/deprovisioning Prometheus instances (via
+[prometheus-operator][]), [OpsGenie][] heartbeats and other associated
+resources. Another component is [Promxy][] which brings all the per-cluster
+shards together and makes them appear as a single API endpoint to the user.
+
+## Testing Goals
+
+We would like the tests to cover the following:
+
+1. Deployment:
+
+  - ensure all components are installable together; this is to catch any issues
+    with Helm charts, etc.
+
+  - ensure that components are installable on all supported infrastructure
+    providers; this is again to catch any issues with Helm charts or templating
+
+2. Provisioning/Deprovisioning operation:
+
+  - ensure that all required resources are created/cleaned-up in reaction to
+    cluster provisioning/deprovisioning respectively; this is to catch bugs and
+    regressions in _prometheus-meta-operator_
+
+  - ensure the required Prometheus instances are provisioned and running
+    correctly; this is to catch issues with configuration for Prometheus that
+    we generate
+
+3. Monitoring system operation:
+
+  - ensure that all built-in/configured targets are configured in provisioned
+    Prometheus instances; this is to catch issues with configuration we
+    generate and catch regressions when monitored components change
+
+  - ensure that provisioned Prometheus instances ingest the metrics we need for
+    alerting; this is to catch issues with configuration and regressions when
+    monitored components change
+
+## Test Infrastructure
+
+Given the above goals, we'd ideally need a system that is able to provision
```

Comment:
Yeah, if this is a hard requirement, we're probably still a few months away from that :D - are there any smaller steps we can take that start us in that direction? For example, does only installing the operator on a small cluster start flushing out a lot of the Helm issues, for example?

Reply:
We discussed it with @TheoBrigitte a little and we thought using "production" test installations like gaia and godsmack could be a good starting point

Reply:
Oh sure, we integrate w/ all the fancy tekton stuff. I hadn't thought about that, nice!

Reply:
Okay, given that, it'd be cool to see some smaller initial steps (e.g: 80/20 rule) - what's the first small thing we can do?

Reply:
TBH, prometheus-meta-operator is an app so I think we should check with ManagedApp teams what is their solution to test app across the different providers, including how they do e2e tests and maybe use what they use?

Reply:
If we only want the easy solution, we can just deploy the app across all 3 providers and then add tests over time

Reply:
Yes we could test on small cluster, but is anything preventing us to test on testing installations ? My thinking is the closer we test on things that match production environment the better.

Reply:
> TBH, prometheus-meta-operator is an app so I think we should check with ManagedApp teams what is their solution to test app across the different providers, including how they do e2e tests and maybe use what they use?

I like that. For deployability let's not re-invent stuff, other teams would have the same issues with their components.

Reply:
yo @cokiengchiara / @piontec  do y'all have any thoughts on testing apps like this?

Reply:
We're not there yet - I mean testing across different providers. We have this on our map, and I hope we can do it soon, but so far we only have a sketch of the test framework we might be using for that. If you're interested, have a look here https://github.com/giantswarm/app-build-suite/tree/master/examples/apps/hello-world-app/tests/abs

The end goal is to have something like that: https://docs.google.com/presentation/d/1EP1TweMtMg-YkTCSgKUU5ec6yP2L2ZteaqD7WC24WH8/edit#slide=id.gafe28231d1_0_165